zend_abstract_interface/symbols/lookup.c

```diff
@@ -0,0 +1,519 @@
+            ce = (zend_class_entry*) scope;
+        break;
+
+        case ZAI_SYMBOL_SCOPE_OBJECT:
+            ce = Z_OBJCE_P((zval*) scope);
+        break;
+    }
+
+    if (!zai_symbol_update(ce ZAI_TSRMLS_CC)) {
+        return NULL;
+    }
+
+#if PHP_VERSION_ID >= 70100
+    zend_class_constant *constant = zai_symbol_lookup_table(&ce->constants_table, name->ptr, name->len, false, true ZAI_TSRMLS_CC);
+
+    if (!constant) {
+        return NULL;
+    }
+
+    return &constant->value;
+#elif PHP_VERSION_ID >= 70000
+    zval *constant = zai_symbol_lookup_table(&ce->constants_table, name->ptr, name->len, false, false ZAI_TSRMLS_CC);
+
+    if (!constant) {
+        return NULL;
+    }
+
+    return constant;
+#else
+    zval **constant = zai_symbol_lookup_table(&ce->constants_table, name->ptr, name->len, false, false ZAI_TSRMLS_CC);
+
+    if (!constant) {
+        return NULL;
+    }
+
+    return *constant;
+#endif
+}
+
+static inline zval* zai_symbol_lookup_constant_impl(
+        zai_symbol_scope_t scope_type, void *scope,
+        zai_string_view *name ZAI_TSRMLS_DC) {
+    switch (scope_type) {
+        case ZAI_SYMBOL_SCOPE_GLOBAL:
+        case ZAI_SYMBOL_SCOPE_NAMESPACE:
+            return zai_symbol_lookup_constant_global(scope_type, scope, name ZAI_TSRMLS_CC);
+
+        case ZAI_SYMBOL_SCOPE_CLASS:
+        case ZAI_SYMBOL_SCOPE_OBJECT:
+            return zai_symbol_lookup_constant_class(scope_type, scope, name ZAI_TSRMLS_CC);
+
+        default:
+            assert(0 && "constant lookup may not be performed in static and frame scopes");
+            return NULL;
+    }
+}
+
+static inline zval* zai_symbol_lookup_property_impl(
+        zai_symbol_scope_t scope_type, void *scope,
+        zai_string_view *name ZAI_TSRMLS_DC) {
+    zend_property_info *info = NULL;
+    zend_class_entry *ce = NULL;
+
+    switch (scope_type) {
+        case ZAI_SYMBOL_SCOPE_OBJECT:
+            ce = Z_OBJCE_P((zval*) scope);
+        break;
+
+        case ZAI_SYMBOL_SCOPE_CLASS:
+            ce = (zend_class_entry*) scope;
+        break;
+
+        default:
+            assert(0 && "property lookup may only be performed in class and object scopes");
+            return NULL;
+    }
+
+    info = zai_symbol_lookup_table(&ce->properties_info, name->ptr, name->len, false, true ZAI_TSRMLS_CC);
+
+    if (!info) {
+        if (scope_type == ZAI_SYMBOL_SCOPE_OBJECT) {
+#if PHP_VERSION_ID < 70000
+            zend_object *obj = zend_object_store_get_object((zval*) scope ZAI_TSRMLS_CC);
+#else
+            zend_object *obj = Z_OBJ_P((zval*) scope);
+#endif
+
+            void *property = zai_symbol_lookup_table(
+                obj->properties,
+                name->ptr, name->len,
+                false, false ZAI_TSRMLS_CC);
+
+#if PHP_VERSION_ID < 70000
+            if (!property) {
+                return NULL;
+            }
+
+            return *(zval**) property;
+#else
+            return (zval*) property;
```

Comment:
why do we deref local frame lookup, but not property?

Reply:
fixed that